### ISSUE TITLE
Add a reactive transaction API

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13502]
An important leftover from jOOQ 3.15's R2DBC support (see https://github.com/jOOQ/jOOQ/issues/11700) is a jOOQ-native transaction API that simplifies transactional logic, which can be cumbersome if done with R2DBC API directly (see also open issues here: https://github.com/jOOQ/jOOQ/issues/11717).

A new API will be added:

- `TransactionalPublishable<T>`, a functional interface implementing the support
- `DSLContext.transactionPublisher(TransactionalPublishable<T>): Publisher<T>`

The `TransactionalPublishable` type looks like this:

```java
@FunctionalInterface
public interface TransactionalPublishable<T> {

    /**
     * Run the transactional code.
     *
     * @param configuration The <code>Configuration</code> in whose context the
     *            transaction is run.
     * @return The outcome of the transaction.
     */
    Publisher<T> run(Configuration configuration);
}
```

Client code could look like this:

```java
Flux<?> r = Flux.from(DSL.using(dbSetup.connectionFactory).transactionPublisher(x -> Flux
    .from(x.dsl().query("insert into t (i, j) values (1, 1)"))
    .thenMany(x.dsl().query("insert into t (i, j) values (2, 1)"))
));
```

nesting transactions with savepoints

```java
Flux<?> r = Flux.from(DSL.using(dbSetup.connectionFactory).transactionPublisher(x -> Flux
    .from(x.dsl().query("insert into t (i, j) values (1, 1)"))

    // Savepoint here
    .thenMany(x.dsl().transactionPublisher(y -> Flux
        .from(x.dsl().query("insert into t (i, j) values (2, 1)"))))
));
```

### Tasks

- [x] Add API
- [x] Add a non-blocking implementation with these features
  - [x] The usual connection auto-closing applies, if necessary
  - [x] Successful completion of the publishable results in a commit
  - [x] Failure produced by the publishable results in a rollback
  - [x] Nesting publishables results in savepoint usage (and rollback roll back to the savepoint)
- [x] Add a blocking compatibility implementation (if a JDBC `Connection` is configured, instead of an R2DBC `Connection`)
- [x] Add documentation

### Out of scope

- `TransactionListener` integration
- `TransactionProvider` integration

### Caveats

- Unlike when returning a `org.jooq.Publisher` (which is both an `org.reactivestreams.Publisher` and a `java.util.concurrent.Flow.Publisher`), or when accepting the two types in API, it will be harder for a functional interface to be able to work with both. It's probably sensible to work with `org.reactivestreams.Publisher` for now.

**Comment by lukaseder** [comment]
Pushed a first draft that was treacherously simple to pull off? Can't be true. Will look into making this functionality pass the TCK and add support for savepoints tomorrow.

**Comment by JohannesZick** [comment]
This looks to me a lot like the Spring API: a lambda that receives a configuration and returns a publisher, and the transaction concludes when the publisher concludes. I like it, it's quite intuitive.

I'm less sure about the savepoints. The simple transaction works well because the one publisher gives a clear spot for start and finish. This can be mapped to start and finish of a transaction. But when we start nesting publishers, it's in general much harder to establish something like an order or just clear before/after distinctions. Things might very well run in parallel, with savepoints being set in a random order.

Also, with the way errors work in a reactive setting, I wonder how to handle rollbacks with other error handling in realistic, complex scenarios. :thinking: 

TLDR, I would personally be fine with the "basic" version for now and leave savepoints for a future release.

**Comment by lukaseder** [comment]
> a lambda that receives a configuration and returns a publisher, and the transaction concludes when the publisher concludes.

It's mostly also consistent with jOOQ's blocking transaction API

> Things might very well run in parallel, with savepoints being set in a random order.

Yes, but that's the user's problem. Users don't *have to* nest transactions, and users don't *have to* leave parallelism to the gods of randomness.

Weird, unwanted parallelism and other scheduling fun is at the core of reactive programming, at least when the RS APIs are used by people who aren't aware of what the various primitives really mean.

The whole point of RS is to enable this parallelism at the risk of using it by accident (probably a lot).

I'm not convinced that jOOQ should help prevent that.

> Also, with the way errors work in a reactive setting, I wonder how to handle rollbacks with other error handling in realistic, complex scenarios. 🤔

It's simple. A `TransactionalPublishable` (name might change) commits (or releases the savepoint) upon completion, and rollbacks (or rollbacks to the savepoint) upon error. This is how jOOQ's blocking transaction API works, and there's no reason to make the non-blocking transaction API work differently.

> I would personally be fine with the "basic" version for now and leave savepoints for a future release.

I don't see why that should be left for a future release. I don't see any other way to design nested transactions. But again, users don't *have to* nest their transactions.

**Comment by JohannesZick** [comment]
True, no one has to do it. After working with Reactor for a while, I've just become hesitant. It's so easy to make mistakes and have unwanted results, that I immediately react to any API that adds more potential for mistakes on top.

In the end, I agree that you cannot stop people from misusing things. But at the same time, I prefer an API that makes it harder. And at first glance, I don't even know how to use savepoints correctly. Then again, I also don't know a better API for them either.

Sorry, I don't even know where I'm going with this. I think I'm just triggered by the insanity of reactive programming.

**Comment by lukaseder** [comment]
> True, no one has to do it. After working with Reactor for a while, I've just become hesitant. It's so easy to make mistakes and have unwanted results, that I immediately react to any API that adds more potential for mistakes on top.

Well, ... Why go reactive in the first place? It's one of the biggest footguns I've ever seen, and if it doesn't really really (really, *REALLY*) add the value to your system that it is capable of, I would 100% recommend against it. I'm very positive that 80% of those folks who want to use reactive with jOOQ don't actually *need* it, they just *want* it (or at some point thought they had wanted it). 😅

Loom will put an end to all of this, eventually. It's ready now, but it will still take 10 years until all the libraries (including JDBC drivers) will be "Loom-capable." Until then, R2DBC and the RS APIs do add some value, but they produce an amount of cognitive friction that will be gone with Loom.

> In the end, I agree that you cannot stop people from misusing things. But at the same time, I prefer an API that makes it harder. And at first glance, I don't even know how to use savepoints correctly. Then again, I also don't know a better API for them either.

I'm not really sure why it's harder. There are only 2 really meaningful transaction propagations (e.g.  in `org.springframework.transaction.annotation.Propagation`):

- `NESTED` (what jOOQ does, by default)
- `REQUIRED` (what you're suggesting)

Some others (e.g. `REQUIRES_NEW`) can be occasionally useful, but certainly not as the default. It's just a matter of what should be the default. I guess Spring defaulted to `REQUIRED` because that's the one that's most widely supported. Not all RDBMS support savepoints, and you need those to implement the `NESTED` semantics. But is `REQUIRED` really simpler? I find it very confusing.

A closure / scope / lambda / whatever you want to call it is a computational unit. It can be executed "top level" or in a "nested" scope. It makes perfect sense to make it transactional (ACID), in case of which whatever it does can either be propagated to the caller atomically, or rolled back atomically. But the nested unit should not roll back anything from outside the unit. Nor should it be exited with an error and weird state. Any caller of that unit can assume that the unit will either not touch the data as it was before the unit was called, or it will leave it in a consistent state. The caller can't and shouldn't clean up anything that was left inconsistently, because it wouldn't know how to.

If that doesn't match the description of the unit, because the unit is just a utility that does a bunch of re-usable, non-transactional DML statements, then it isn't going to be a transactional unit. But if the unit is transactional, then why this wishy-washy `REQUIRED` behaviour? This makes me think that `REQUIRED` is the default in Spring for merely incidental reasons, because in Spring, everything tends to be `public`, and every sloppy programmer starts calling what should be internal methods from some entirely different context, and suddenly, all the code has to be transactional capable, even if it shouldn't be.

This isn't the case (or at least less so) with jOOQ's programmatic transaction API, because it's much less incidental, and much more explicit.

> Sorry, I don't even know where I'm going with this. I think I'm just triggered by the insanity of reactive programming.

Well, if you're not 100% certain that you *need* your application to be reactive, then why not just switch back to using JDBC? At least, from the jOOQ side, all of your code can remain almost the same :)

**Comment by JohannesZick** [comment]
Well, glad to see we are on the same page for reactive code in general. Using it for a while makes me appreciate Kotlin Coroutines.

I'll be sure to give this a try when it's available, so I can leave some more useful feedback. ;)

**Comment by lukaseder** [comment]
> Well, glad to see we are on the same page for reactive code in general. Using it for a while makes me appreciate Kotlin Coroutines.

Yeah, I have to play around with an integration once. I think they solve some of the flow related craziness that happens otherwise with RS APIs, such that the code at least *looks* sequential.

In case of which, the nesting becomes quite apparent from the flow, I think? And parallel accidents are unlikely?

> I'll be sure to give this a try when it's available, so I can leave some more useful feedback. ;)

It is definitely useful as a warning for others who are thinking about going reactive :)

**Comment by lukaseder** [comment]
> Using it for a while makes me appreciate Kotlin Coroutines.

@JohannesZick You don't happen to have a small sample snippet to show how jOOQ/R2DBC integrates with Kotlin coroutines, for you?

**Comment by JohannesZick** [comment]
> @JohannesZick You don't happen to have a small sample snippet to show how jOOQ/R2DBC integrates with Kotlin coroutines, for you?

Not exactly. I use R2DBC and jOOQ in Spring and Reactor at work. Would not recommend it unless you really need the whole "process terrabytes of data in lazy streams" aspect. Just way too much unnecessary complications. I use JDBC and jOOQ in a personal project with Kotlin Coroutines. I mostly write extension functions on the DSL like this:

```kotlin
suspend fun DSLContext.replaceStuff(parentId: Long, stuff: List<Stuff>) = withContext(Dispatchers.IO) {
    transaction { configuration: Configuration ->
        configuration.dsl().deleteFrom(STUFF)
            .where(STUFF.PARENT_ID.eq(parentId))
            .execute()
        configuration.dsl().batchStore(
            stuff.map {
                StuffRecord()
                    .with(STUFF.NAME, it.name)
                    .with(STUFF.VALUE, it.value)
                    .with(STUFF.PARENT_ID, parentId)
            }
        )
            .execute()
    }
}
```

Coroutines allow for easy parallelism, and in general code might run on a different thread after the coroutine was paused. But on the other hand, it's easy to run blocking code like this. The IO dispatcher comes OOTB for exactly this purpose. Since I don't interact with coroutines inside the block here, I know it all runs in one thread and I have zero issues using transactions. At the same time, calling this won't change threads of anything else.

If I was using R2DBC, I would use the library to bridge reactive streams to coroutines and simply call `await()` or something like it on all queries. I have not used this with transactions yet. I think jOOQ doesn't use ThreadLocals, so I would not worry personally. Otherwise, I imagine wrapping the transaction in a Dispatcher that uses just one thread should do the trick. No guarantees though.

**Comment by JohannesZick** [comment]
Thanks! I'm looking forward to playing around with this. No promises on a timeline for feedback though. :sweat_smile:

**Comment by lukaseder** [comment]
Done. I'm really surprised how easy this turned out to be, given how hard it was to correctly implement the `Publisher` API and pass TCK.
